### PR TITLE
[#888] Adding multiple address fields functionality

### LINKED ISSUE #888: In admin the field "number of lines in the street address" is ignored
**Description**:
The field "number of lines in the street address" does not affect the number of fields for the street address in account settings or checkout page.

**Expected behavior**:

Changing the value in admin in the field "number of lines in the street address" should change the number of fields for a street address in account settings and checkout.

**Screenshots, Video, Logs**

image

image

### PR DESCRIPTION
**Pull request by MaksimGritcenko**
fixes #888

### CODE REVIEW

**Comment by MaksimGritcenko** [comment]
BE part for this issue scandipwa/store-graphql#6

**Review by yeegor — CHANGES_REQUESTED** [review]

File: src/app/component/CheckoutBilling/CheckoutBilling.container.js [hunk 1]
```diff
@@ -121,8 +123,10 @@ export class CheckoutBillingContainer extends PureComponent {
     }
 
     onBillingSuccess(fields, asyncData) {
-        const { savePaymentInformation } = this.props;
-        const address = this._getAddress(fields);
+        const { savePaymentInformation, addressLinesQty } = this.props;
+        const address = this._getAddress(addressLinesQty > 1
```
[the remark this answers is not in the document]
Comment (MaksimGritcenko) [reply src/app/component/CheckoutBilling/CheckoutBilling.container.js:127]: I don`t handle this comparison in this function.

File: src/app/util/Address/index.js [hunk 2]
```diff
@@ -49,3 +49,26 @@ export const trimAddressFields = (fields) => {
...
+/** resets street in checkout/billing/myAccoutAddress form fields object */
+/** @namespace Util/Address/setAddressesInFormObject */
+export const setAddressesInFormObject = (fields, numberOfLines) => {
+    const addressKeys = new Array(numberOfLines)
+        .fill('')
+        .map((_, index) => `street${index}`);
+
+    const addressValues = addressKeys.map((key) => fields[key]);
+
+    const newFields = Object.keys(fields)
```
Comment (yeegor) [inline src/app/util/Address/index.js:62]: Please comment this

File: src/app/component/CheckoutBilling/CheckoutBilling.container.js [hunk 3]
```diff
@@ -121,8 +123,10 @@ export class CheckoutBillingContainer extends PureComponent {
...
+            ? setAddressesInFormObject(fields, addressLinesQty)
```
Comment (yeegor) [inline src/app/component/CheckoutBilling/CheckoutBilling.container.js:128]: I believe this logic should be a part of _getAddress function

File: src/app/component/CheckoutShipping/CheckoutShipping.container.js [hunk 4]
```diff
@@ -73,7 +75,7 @@ export class CheckoutShippingContainer extends PureComponent {
 
         const shippingAddress = selectedCustomerAddressId
             ? this._getAddressById(selectedCustomerAddressId)
-            : trimAddressFields(fields);
+            : trimAddressFields(addressLinesQty > 1 ? setAddressesInFormObject(fields, addressLinesQty) : fields);
```
Comment (yeegor) [inline src/app/component/CheckoutShipping/CheckoutShipping.container.js:78]: This is poorly readable, please split the function call with argument selection

File: src/app/component/MyAccountAddressForm/MyAccountAddressForm.component.js [hunk 5]
```diff
@@ -93,10 +98,37 @@ export class MyAccountAddressForm extends FieldForm {
...
 
+    getStreetFields(label, value) {
+        return {
+            label: __(label),
+            value,
+            validation: ['notEmpty']
+        };
+    }
+
+    getAddressFields() {
```
Comment (yeegor) [inline src/app/component/MyAccountAddressForm/MyAccountAddressForm.component.js:109]: Please comment on why this function is necessary

**Review by yeegor — CHANGES_REQUESTED** [review]

File: src/app/component/CheckoutShipping/CheckoutShipping.container.js [hunk 6]
```diff
@@ -64,16 +66,20 @@ export class CheckoutShippingContainer extends PureComponent {
...
     onShippingSuccess(fields) {
-        const { saveAddressInformation } = this.props;
+        const { saveAddressInformation, addressLinesQty } = this.props;
 
         const {
             selectedCustomerAddressId,
             selectedShippingMethod
         } = this.state;
 
+        const formFields = addressLinesQty > 1
```
Comment (yeegor) [inline src/app/component/CheckoutShipping/CheckoutShipping.container.js:76]: I believe this logic can be hidden in "getFormFields" function with the only responsibility of handling this logic.
Please use this approach in order to get rid of the ternary operator and to make code more readable.

File: src/app/component/CheckoutBilling/CheckoutBilling.container.js [hunk 7]
```diff
@@ -183,19 +186,23 @@ export class CheckoutBillingContainer extends PureComponent {
...
-        const { shippingAddress } = this.props;
+        const { addressLinesQty, shippingAddress } = this.props;
 
         const {
             isSameAsShipping,
             selectedCustomerAddressId
         } = this.state;
 
+        const formFields = addressLinesQty > 1
+            ? setAddressesInFormObject(fields, addressLinesQty)
```
Comment (yeegor) [inline src/app/component/CheckoutBilling/CheckoutBilling.container.js:197]: Read the comment below about the same logic.

**Review by yeegor — CHANGES_REQUESTED** [review]
Overall OK, but please introduce changes corresponding to the comments I have left.

File: src/app/util/Address/index.js [hunk 8]
```diff
@@ -49,3 +49,28 @@ export const trimAddressFields = (fields) => {
...
+export const setAddressesInFormObject = (fields, numberOfLines) => {
+    const addressKeys = new Array(numberOfLines)
+        .fill('')
+        .map((_, index) => `street${index}`);
+
+    const addressValues = addressKeys.map((key) => fields[key]);
+
+    // reseting multiple form object street entries into
+    // one entry "street", equal to array of field values
+    const newFields = Object.keys(fields)
```
[the remark this answers is not in the document]
Comment (MaksimGritcenko) [reply src/app/util/Address/index.js:64]: but there are related fields. They just need to get transformed into a single array

**Pull request merged**